hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieDefaultTimeline.java

```diff
@@ -399,4 +399,19 @@ public boolean isEmpty(HoodieInstant instant) {
   public String toString() {
     return this.getClass().getName() + ": " + instants.stream().map(Object::toString).collect(Collectors.joining(","));
   }
+
+  /**
+   * Merge this timeline with the given timeline.
+   */
+  public HoodieDefaultTimeline mergeTimeline(HoodieDefaultTimeline timeline) {
+    Stream<HoodieInstant> instantStream = Stream.concat(instants.stream(), timeline.getInstants()).sorted();
+    Function<HoodieInstant, Option<byte[]>> details = instant -> {
+      if (instants.stream().anyMatch(i -> i.equals(instant))) {
+        return this.getInstantDetails(instant);
+      } else {
+        return timeline.getInstantDetails(instant);
+      }
```

Comment:
I went through the code. The archived timeline details are computed just once. But active tiimeline instant details are re-computed. Typically, for older tables, archive timeline is much bigger than active timeline so the amortized cost should be within bounds. 

To optimize for active timeline, we need a map/cache of instant details in `HoodieActiveTimeline` and then in `HoodieActiveTimeline#getInstantDetails`, we have to check this cache before reading from file. I can take it up in a separate PR. It might be beneficial at other places too, irrespective of this patch.